pandas/core/nanops.py

```diff
@@ -421,8 +421,44 @@ def nanany(values, axis=None, skipna: bool = True, mask=None):
     >>> nanops.nanany(s)
     False
     """
-    values, _, _, _, _ = _get_values(values, skipna, fill_value=False, mask=mask)
-    return values.any(axis)
+    values, _, dtype, _, _ = _get_values(values, skipna, fill_value=False, mask=mask)
+
+    # GH #12863
+    # Checking if the `axis` is None because numpy
+    # doesn't handle ``any`` and ``all`` on
+    # object arrays correctly. see
+    # https://github.com/numpy/numpy/issues/4352
+
+    # TODO: Find a less code-smelly way of doing this
+    if is_object_dtype(dtype) and axis is None:
+        output = values.any()
+    else:
+        output = values.any(axis)
+
+    if isinstance(output, bool):
```

Comment:
when is this NOT true?

Reply:
That's exactly the bug, for example (from the original issue).

master:
```python
>>> import pandas as pd
>>> s = pd.Series(index=range(5), data=['a', 'b', 'c', 'd', 'e'], dtype=object)
>>> s.any()
'a'
>>> s.all()
'e'
```

This branch (with brute force function):
```python
>>> import pandas as pd
>>> s = pd.Series(index=range(5), data=['a', 'b', 'c', 'd', 'e'], dtype=object)
>>> s.any()
True
>>> s.all()
True
```